src/wasm-interpreter.h

```diff
@@ -902,286 +909,369 @@ class ModuleInstanceBase {
-                    << function->params.size() << " parameters, got "
-                    << arguments.size() << " arguments." << std::endl;
-          WASM_UNREACHABLE();
-        }
-        locals.resize(function->getNumLocals());
-        for (size_t i = 0; i < function->getNumLocals(); i++) {
-          if (i < arguments.size()) {
-            assert(function->isParam(i));
-            if (function->params[i] != arguments[i].type) {
-              std::cerr << "Function `" << function->name << "` expects type "
-                        << printType(function->params[i])
-                        << " for parameter " << i << ", got "
-                        << printType(arguments[i].type) << "." << std::endl;
-              WASM_UNREACHABLE();
-            }
-            locals[i] = arguments[i];
-          } else {
-            assert(function->isVar(i));
-            locals[i].type = function->getLocalType(i);
-          }
-        }
-      }
-    };
+    Const offset;
+    offset.value = Literal(uint32_t(0));
+    offset.finalize();
 
-    // Executes expressions with concrete runtime info, the function and module at runtime
-    class RuntimeExpressionRunner : public ExpressionRunner<RuntimeExpressionRunner> {
-      ModuleInstanceBase& instance;
-      FunctionScope& scope;
-
-    public:
-      RuntimeExpressionRunner(ModuleInstanceBase& instance, FunctionScope& scope) : instance(instance), scope(scope) {}
-
-      Flow generateArguments(const ExpressionList& operands, LiteralList& arguments) {
-        NOTE_ENTER_("generateArguments");
-        arguments.reserve(operands.size());
-        for (auto expression : operands) {
-          Flow flow = this->visit(expression);
-          if (flow.breaking()) return flow;
-          NOTE_EVAL1(flow.value);
-          arguments.push_back(flow.value);
-        }
-        return Flow();
+    // apply active memory segments
+    for (size_t i = 0, e = wasm.memory.segments.size(); i < e; ++i) {
+      Memory::Segment& segment = wasm.memory.segments[i];
+      if (segment.isPassive) {
+        continue;
       }
 
-      Flow visitCall(Call *curr) {
-        NOTE_ENTER("Call");
-        NOTE_NAME(curr->target);
-        LiteralList arguments;
-        Flow flow = generateArguments(curr->operands, arguments);
-        if (flow.breaking()) return flow;
-        auto* func = instance.wasm.getFunction(curr->target);
-        Flow ret;
-        if (func->imported()) {
-          ret = instance.externalInterface->callImport(func, arguments);
+      Const size;
+      size.value = Literal(uint32_t(segment.data.size()));
+      size.finalize();
+
+      MemoryInit init;
+      init.segment = i;
+      init.dest = segment.offset;
+      init.offset = &offset;
+      init.size = &size;
+      init.finalize();
+
+      DataDrop drop;
+      drop.segment = segment.index;
+      drop.finalize();
+
+      Function initializer;
+      initializer.body = builder.blockify(&init, &drop);
```

Comment:
Unfortunately there's no way to create a `Block` without involving an `ArenaAllocator` as far as I can tell. It didn't seem worth creating a new `ArenaAllocator` for these temporary allocations.

Reply:
Agreed, but please add a comment (maybe on the `builder` creation).